### DISCUSSION TITLE
map downloads are extremely slow

### DISCUSSION THREAD

**Discussion by hlinden** [osmandapp/OsmAnd#12898] [General]
I am currently trying to download some standard maps for my country and I'm seeing 5% of 600MB after half an hour or so. That translates to about 15kB/s. I have switched from my local network to 4G to make sure that nothing about the routing or peering to the content host is wonky, but the situation is not improving.

I get that content hosting is expensive. I'm already subscribed to OSMAnd Live, but I'd be happy to throw a few more Euros per month to you guys if that means I can download maps in a reasonable time.

In the meantime, is there any other way to resolve this?

**Comment by vshcherb** [comment]
Well, we don't exactly why it is slow sometimes. Please double check here https://download.osmand.net/list

**Reply by hlinden** [reply]
I saw that that site redirects to dlSOMETHING.osmand.net, so I whipped this oneliner up for benchmarking:

```
for host in $(for i in {1..20}; do host dl$i.osmand.net | grep -v NXDOMAIN | awk '{print $1}'; done); do echo "Downloading from $host."; curl --silent -w '%{speed_download} bytes/s average speed.\n' -m30 "https://$host/download?standard=yes&file=Germany_hessen_europe_2.obf.zip" -o /dev/null; done
```


...and it returned for me:

```
Downloading from dl2.osmand.net.
12989607 bytes/s average speed.
Downloading from dl7.osmand.net.
69350 bytes/s average speed.
Downloading from dl8.osmand.net.
55696 bytes/s average speed.
Downloading from dl9.osmand.net.
83001 bytes/s average speed.
```

I've got a 100MBit/s VDSL line, so basically dl2 maxed that out and all the other ones were extremely slow.

I gave it another try on a machine hosted at Netcup in Germany and all the mirrors were properly fast:

```
Downloading from dl2.osmand.net.
49980146.000 bytes/s average speed.
Downloading from dl7.osmand.net.
75597585.000 bytes/s average speed.
Downloading from dl8.osmand.net.
93213188.000 bytes/s average speed.
Downloading from dl9.osmand.net.
78956254.000 bytes/s average speed.
```

So, it boils down to: Would it be possible to redirect Telekom Germany customers to dl2 only?

**Reply by vshcherb** [reply]
That's interesting and strange... let's if it's not fixed in couple days.

**Reply by hlinden** [reply]
It's still bad.

```
Downloading from dl2.osmand.net.
12795405 bytes/s average speed.
Downloading from dl7.osmand.net.
267603 bytes/s average speed.
Downloading from dl8.osmand.net.
259411 bytes/s average speed.
Downloading from dl9.osmand.net.
235380 bytes/s average speed.
```

I guess, the problem is that, while `dl2` ist hosted at Hetzner, Germany, `dl7`, `dl8` and `dl9` are hosted at poneytelecom aka AS12876 aka Scaleway. The peering from Telekom Germany to them (via Level 3) is just crap and during even slightly busy times of the day the bandwidth drops beneath 5 MBit/s.

Their bandwidth test at http://ping.online.net/ shows results consistent with the rates received from the OSMand servers - it definitely is a problem at the hoster level and the best way would be to redirect users from Telekom Germany to `dl2` for the time being.


I have no idea how one would do that in a reasonable manner, maybe the list of the announced subnets from AS 3320 is a good start: https://bgp.he.net/AS3320#_prefixes

**Reply by vshcherb** [reply]
Today the problem was with Scaleway actually, they are replacing the routing. Will test tomorrow.

**Reply by hlinden** [reply]
The problem seems to persist. I just ran the benchmark from a wired machine on a Telekom Germany VDSL 100 connection and from another machine connected to my phone which uses T-Mobile LTE:

VDSL 100, Telekom Germany
```
Downloading from dl2.osmand.net.
9520502 bytes/s average speed.
Downloading from dl7.osmand.net.
65527 bytes/s average speed.
Downloading from dl8.osmand.net.
54058 bytes/s average speed.
Downloading from dl9.osmand.net.
62248 bytes/s average speed.
```

LTE, T-Mobile Germany
```
Downloading from dl2.osmand.net.
6052986 bytes/s average speed.
Downloading from dl7.osmand.net.
35489 bytes/s average speed.
Downloading from dl8.osmand.net.
33850 bytes/s average speed.
Downloading from dl9.osmand.net.
31666 bytes/s average speed.
```

some VM hosted at Netcup for comparison
```
Downloading from dl2.osmand.net.
43133593.000 bytes/s average speed.
Downloading from dl7.osmand.net.
45668378.000 bytes/s average speed.
Downloading from dl8.osmand.net.
54580585.000 bytes/s average speed.
Downloading from dl9.osmand.net.
84099948.000 bytes/s average speed.
```

This would make it impossible to download new maps for a reasonable region size even overnight. As the peering from Telekom to Scaleway still seems to be the problem, redirecting users from Telekom Germany to `dl2` would probably be a good idea.

**Reply by vshcherb** [reply]
We're planning to setup dynamic routing at least for Germany to dl2. Do you have any simple detection algorithm idea?

**Reply by hlinden** [reply]
As the problem seems to be related only to Telekom Germany (which includes T-Mobile) and their peering, I'd probably pull a list of their announced IPv4 prefixes, e.g. from `https://stat.ripe.net/data/announced-prefixes/data.json?resource=AS3320` on a regular basis and use the list in your script at `https://download.osmand.net/download?` to match and redirect clients from those prefixes to `dl2` only. This should cover the current problem without any external lookups (reverse DNS, whatever...) at runtime and could be easily expanded to fix similar situations with other ISPs.

**Reply by vshcherb** [reply]
https://github.com/osmandapp/OsmAnd/issues/13032

**Comment by da-alb** [comment]
The issue persists even on Android 14 with Osmand 4.7.17. Downloading directly from the link above there are no issues.
